driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java

```diff
@@ -130,41 +120,56 @@ public boolean isClosed() {
         return closed;
     }
 
-    ServerCursor getServerCursor() {
-        return cursor;
+    private boolean getAndSetClosed() {
+        boolean oldClosed;
+        synchronized (lock) {
+            oldClosed = closed;
+            closed = true;
+        }
+        return oldClosed;
+    }
+
+    private ServerCursor getAndSetCursor(final ServerCursor cursor) {
+        ServerCursor oldCursor;
+        synchronized (lock) {
+            oldCursor = this.cursor;
+            this.cursor = cursor;
+        }
+        return oldCursor;
     }
 
     private boolean limitReached() {
         return Math.abs(limit) != 0 && count >= Math.abs(limit);
     }
 
-    private void getMore(final SingleResultCallback<List<T>> callback) {
+    private void getMore(final ServerCursor cursor, final SingleResultCallback<List<T>> callback) {
         connectionSource.getConnection(new SingleResultCallback<AsyncConnection>() {
```

Comment:
The race I was referring to is between next() and close():

T1: next() gets non-null cursor
T2: close() sets cursor to null
T3: close() kills the cursor
T4: close() (asynchronously) releases connectionSource
T5: next() calls connectionSource.getConnection, but refcount is 0

The most recent commit has a similar race condition.

T1: next() sets cursor to null, and gets a non-null cursor
T2: close() sets cursor to null, and gets a null cursor
T3: close() (synchronously) releases connectionSource since cursor is null
T4: next() calls connectionSource.getConnection, but refcount is 0

Is also changes the behavior of the cursor, as now there is no way to kill a cursor while a getMore is in flight.  And once the getMore completes, the cursor won't be closed once it's done, potentially leaving an open cursor on the server.  I don't think we should make this change.